[inductor] For View.create(x, sizes) call realize_input() instead of realize() when handling unbacked symints

ColinPeppler opened the pull request:
# Context
Let's say we do `View.create(x, sizes)` where `x` is a `SliceView` and `sizes` contains unbacked symints e.g. `sizes = [i14, 256]`. Then, this we'll run ([this code](https://github.com/pytorch/pytorch/blob/7e37f63e5e77a217a08d9261af49e63ddb51d306/torch/_inductor/ir.py#L2058-L2071)) where we.
1. Call `x.realize()` -- SliceView(Pointwise) -> SliceView(ComputedBuffer).
2. Retrieve storage & layout via `as_storage_and_layout(x)`
3. Calculate `new_layout` based off layout & `new_sizes`
3. `return ReinterpretView(storage, new_layout)`
However, (2) will raise `NotImplementedError` ([see](https://github.com/pytorch/pytorch/blob/7e37f63e5e77a217a08d9261af49e63ddb51d306/torch/_inductor/ir.py#L1704-L1731)) since `x` is a `SliceView` and that isn't supported.

Thus, I tried adding support for `SliceView` in `as_storage_and_layout`. This worked for my case, but if instead `sizes` had backed symints e.g. `sizes=[s0, 256]` then some benchmarked models lost accuracy.
```
    if isinstance(x, SliceView):
        return as_storage_and_layout(
            x.data,
            freeze=freeze,
            want_contiguous=want_contiguous,
            stride_order=stride_order,
        )
```

So instead of the above, I tried unwrapping the `SliceView` via `x = x.unwrap_view()`. This works for my usecase and passes CI but I'm not entirely sure why. If we unwrap our `SliceView` and create a `ReinterpretView`, I'd assume we'd lose the reindexer from `SliceView`. ~~But maybe we can re-create the same indexing from the `ReinterpretView`'s strides?~~ edit: we do lose vital information (like offset) when you release your `SliceView` and create a `ReinterpretView` so that's a no-go.

Moving onto the final version of this PR. We call `ExternKernel.realize_input()` (feels a bit weird to use `ExternKernel` but it's exactly what I need). It will go ahead and handle our `SliceView` case ([see](https://github.com/pytorch/pytorch/blob/a468b9fbdf75b2da128e5fac4284ebbb981cd31a/torch/_inductor/ir.py#L3733-L3739)) by converting it to a `ReinterpretView` with the correct offset.


# Test
```
$ python test/inductor/test_unbacked_symints.py
..
----------------------------------------------------------------------
Ran 10 tests in 20.813s

OK
```

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #117013



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

ColinPeppler committed:
[inductor] skip if not reinterpret_view and unbacked symint

[ghstack-poisoned]

ColinPeppler commented:
Need to write unit test but requesting review for some thoughts

ezyang commented:
I don't know enough about this code to evaluate this change, but you should add a test ;)

jansel reviewed (changes requested):
Add a test please

ColinPeppler committed:
Update on "[inductor] skip if not reinterpret_view and unbacked symint"

# Context
Let's say we do `View.create(x, sizes)` where `x` is a `SliceView` and `sizes` contains unbacked symints e.g. `sizes = [i14, 256]`. Then, this we'll run ([this code](https://github.com/pytorch/pytorch/blob/7e37f63e5e77a217a08d9261af49e63ddb51d306/torch/_inductor/ir.py#L2058-L2071)) where we.
1. Call `x.realize()` -- SliceView(Pointwise) -> SliceView(ComputedBuffer).
2. Retrieve storage & layout via `as_storage_and_layout(x)`
3. Calculate `new_layout` based off layout & `new_sizes`
3. `return ReinterpretView(storage, new_layout)`
However, (2) will raise `NotImplementedError` ([see](https://github.com/pytorch/pytorch/blob/7e37f63e5e77a217a08d9261af49e63ddb51d306/torch/_inductor/ir.py#L1704-L1731)) since `x` is a `SliceView` and that isn't supported.

Thus, I tried adding support for `SliceView` in `as_storage_and_layout`. This worked for my case, but if instead `sizes` had backed symints e.g. `sizes=[s0, 256]` then some benchmarked models lost accuracy.
```
    if isinstance(x, SliceView):
        return as_storage_and_layout(
            x.data,
            freeze=freeze,
            want_contiguous=want_contiguous,
            stride_order=stride_order,
        )
```

So instead of the above, I tried unwrapping the `SliceView` in this PR. This works for my usecase and passes CI but I'm not entirely sure why. If we unwrap our `SliceView` and create a `ReinterpretView`, I'd assume we'd lose the reindexer from `SliceView`. But maybe we can re-create the same indexing from the `ReinterpretView`'s strides?




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ezyang reviewed (approved):
@jansel having recently fixed a similar bug in this vein (https://github.com/pytorch/pytorch/pull/117444) it makes sense to me why this would fix a bug, but what I am less sure about is whether or not this is indicative of some sort of systemic problem in the rest of inductor: the `x.realize()` idiom is *very* common in our lowerings. In my other PR, I reasoned correctness through ExternKernelAlloc, where it seemed to me that you needed to do the "harder" realize to materialize a real tensor. But I am less sure about the generalizable lesson here.

I'll go ahead and stamp this for now

ezyang commented on test/inductor/test_unbacked_symints.py:
```diff
@@ -80,6 +80,23 @@ def fn(x, y):
+
+        example_inputs = (torch.randn(1, 1, 1, 1).cuda(),)
+
+        with dynamo_config.patch({"capture_dynamic_output_shape_ops": True}):
```
nit: you can conveniently use this as a decorator on the test function!

jansel commented:
> @jansel having recently fixed a similar bug in this vein (#117444) it makes sense to me why this would fix a bug, but what I am less sure about is whether or not this is indicative of some sort of systemic problem in the rest of inductor: the `x.realize()` idiom is _very_ common in our lowerings. In my other PR, I reasoned correctness through ExternKernelAlloc, where it seemed to me that you needed to do the "harder" realize to materialize a real tensor. But I am less sure about the generalizable lesson here.

I think we already have that "harder" realize that requires stride orders to be the same.  `require_stride_order` -- it comes with a perf cost though since it could introduce a copy.

ColinPeppler committed:
Update on "[inductor] For View.create(x, sizes) call realize_input() instead of realize() when handling unbacked symints"

# Context
Let's say we do `View.create(x, sizes)` where `x` is a `SliceView` and `sizes` contains unbacked symints e.g. `sizes = [i14, 256]`. Then, this we'll run ([this code](https://github.com/pytorch/pytorch/blob/7e37f63e5e77a217a08d9261af49e63ddb51d306/torch/_inductor/ir.py#L2058-L2071)) where we.
1. Call `x.realize()` -- SliceView(Pointwise) -> SliceView(ComputedBuffer).
2. Retrieve storage & layout via `as_storage_and_layout(x)`
3. Calculate `new_layout` based off layout & `new_sizes`
3. `return ReinterpretView(storage, new_layout)`
However, (2) will raise `NotImplementedError` ([see](https://github.com/pytorch/pytorch/blob/7e37f63e5e77a217a08d9261af49e63ddb51d306/torch/_inductor/ir.py#L1704-L1731)) since `x` is a `SliceView` and that isn't supported.

Thus, I tried adding support for `SliceView` in `as_storage_and_layout`. This worked for my case, but if instead `sizes` had backed symints e.g. `sizes=[s0, 256]` then some benchmarked models lost accuracy.
```
    if isinstance(x, SliceView):
        return as_storage_and_layout(
            x.data,
            freeze=freeze,
            want_contiguous=want_contiguous,
            stride_order=stride_order,
        )
```

So instead of the above, I tried unwrapping the `SliceView` via `x = x.unwrap_view()`. This works for my usecase and passes CI but I'm not entirely sure why. If we unwrap our `SliceView` and create a `ReinterpretView`, I'd assume we'd lose the reindexer from `SliceView`. ~~But maybe we can re-create the same indexing from the `ReinterpretView`'s strides?~~ edit: we do lose vital information (like offset) when you release your `SliceView` and create a `ReinterpretView` so that's a no-go.

Moving onto the final version of this PR. We call `ExternKernel.realize_input()` (feels a bit weird to use `ExternKernel` but it's exactly what I need). It will go ahead and handle our `SliceView` case ([see](https://github.com/pytorch/pytorch/blob/a468b9fbdf75b2da128e5fac4284ebbb981cd31a/torch/_inductor/ir.py#L3733-L3739)) by converting it to a `ReinterpretView` with the correct offset.





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ColinPeppler commented:
@pytorchbot merge

ColinPeppler commented on test/inductor/test_unbacked_symints.py:
```diff
@@ -80,6 +80,23 @@ def fn(x, y):
+
 
 if __name__ == "__main__":
     from torch._dynamo.test_case import run_tests
```
Whoops didn't see this before attempting to merge, I'll quickly test this.

ColinPeppler committed:
Update on "[inductor] For View.create(x, sizes) call realize_input() instead of realize() when handling unbacked symints"

# Context
Let's say we do `View.create(x, sizes)` where `x` is a `SliceView` and `sizes` contains unbacked symints e.g. `sizes = [i14, 256]`. Then, this we'll run ([this code](https://github.com/pytorch/pytorch/blob/7e37f63e5e77a217a08d9261af49e63ddb51d306/torch/_inductor/ir.py#L2058-L2071)) where we.
1. Call `x.realize()` -- SliceView(Pointwise) -> SliceView(ComputedBuffer).
2. Retrieve storage & layout via `as_storage_and_layout(x)`
3. Calculate `new_layout` based off layout & `new_sizes`
3. `return ReinterpretView(storage, new_layout)`
However, (2) will raise `NotImplementedError` ([see](https://github.com/pytorch/pytorch/blob/7e37f63e5e77a217a08d9261af49e63ddb51d306/torch/_inductor/ir.py#L1704-L1731)) since `x` is a `SliceView` and that isn't supported.

Thus, I tried adding support for `SliceView` in `as_storage_and_layout`. This worked for my case, but if instead `sizes` had backed symints e.g. `sizes=[s0, 256]` then some benchmarked models lost accuracy.
```
    if isinstance(x, SliceView):
        return as_storage_and_layout(
            x.data,
            freeze=freeze,
            want_contiguous=want_contiguous,
            stride_order=stride_order,
        )
```

So instead of the above, I tried unwrapping the `SliceView` via `x = x.unwrap_view()`. This works for my usecase and passes CI but I'm not entirely sure why. If we unwrap our `SliceView` and create a `ReinterpretView`, I'd assume we'd lose the reindexer from `SliceView`. ~~But maybe we can re-create the same indexing from the `ReinterpretView`'s strides?~~ edit: we do lose vital information (like offset) when you release your `SliceView` and create a `ReinterpretView` so that's a no-go.

Moving onto the final version of this PR. We call `ExternKernel.realize_input()` (feels a bit weird to use `ExternKernel` but it's exactly what I need). It will go ahead and handle our `SliceView` case ([see](https://github.com/pytorch/pytorch/blob/a468b9fbdf75b2da128e5fac4284ebbb981cd31a/torch/_inductor/ir.py#L3733-L3739)) by converting it to a `ReinterpretView` with the correct offset.





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ColinPeppler commented:
Included both cuda & cpu paths for test_unbacked_symints.py

ColinPeppler committed:
Update on "[inductor] For View.create(x, sizes) call realize_input() instead of realize() when handling unbacked symints"

# Context
Let's say we do `View.create(x, sizes)` where `x` is a `SliceView` and `sizes` contains unbacked symints e.g. `sizes = [i14, 256]`. Then, this we'll run ([this code](https://github.com/pytorch/pytorch/blob/7e37f63e5e77a217a08d9261af49e63ddb51d306/torch/_inductor/ir.py#L2058-L2071)) where we.
1. Call `x.realize()` -- SliceView(Pointwise) -> SliceView(ComputedBuffer).
2. Retrieve storage & layout via `as_storage_and_layout(x)`
3. Calculate `new_layout` based off layout & `new_sizes`
3. `return ReinterpretView(storage, new_layout)`
However, (2) will raise `NotImplementedError` ([see](https://github.com/pytorch/pytorch/blob/7e37f63e5e77a217a08d9261af49e63ddb51d306/torch/_inductor/ir.py#L1704-L1731)) since `x` is a `SliceView` and that isn't supported.

Thus, I tried adding support for `SliceView` in `as_storage_and_layout`. This worked for my case, but if instead `sizes` had backed symints e.g. `sizes=[s0, 256]` then some benchmarked models lost accuracy.
```
    if isinstance(x, SliceView):
        return as_storage_and_layout(
            x.data,
            freeze=freeze,
            want_contiguous=want_contiguous,
            stride_order=stride_order,
        )
```

So instead of the above, I tried unwrapping the `SliceView` via `x = x.unwrap_view()`. This works for my usecase and passes CI but I'm not entirely sure why. If we unwrap our `SliceView` and create a `ReinterpretView`, I'd assume we'd lose the reindexer from `SliceView`. ~~But maybe we can re-create the same indexing from the `ReinterpretView`'s strides?~~ edit: we do lose vital information (like offset) when you release your `SliceView` and create a `ReinterpretView` so that's a no-go.

Moving onto the final version of this PR. We call `ExternKernel.realize_input()` (feels a bit weird to use `ExternKernel` but it's exactly what I need). It will go ahead and handle our `SliceView` case ([see](https://github.com/pytorch/pytorch/blob/a468b9fbdf75b2da128e5fac4284ebbb981cd31a/torch/_inductor/ir.py#L3733-L3739)) by converting it to a `ReinterpretView` with the correct offset.


# Test
```
$ python test/inductor/test_unbacked_symints.py
..
----------------------------------------------------------------------
Ran 10 tests in 20.813s

OK
```




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ezyang commented:
Thanks!!

ColinPeppler committed:
Update on "[inductor] For View.create(x, sizes) call realize_input() instead of realize() when handling unbacked symints"

# Context
Let's say we do `View.create(x, sizes)` where `x` is a `SliceView` and `sizes` contains unbacked symints e.g. `sizes = [i14, 256]`. Then, this we'll run ([this code](https://github.com/pytorch/pytorch/blob/7e37f63e5e77a217a08d9261af49e63ddb51d306/torch/_inductor/ir.py#L2058-L2071)) where we.
1. Call `x.realize()` -- SliceView(Pointwise) -> SliceView(ComputedBuffer).
2. Retrieve storage & layout via `as_storage_and_layout(x)`
3. Calculate `new_layout` based off layout & `new_sizes`
3. `return ReinterpretView(storage, new_layout)`
However, (2) will raise `NotImplementedError` ([see](https://github.com/pytorch/pytorch/blob/7e37f63e5e77a217a08d9261af49e63ddb51d306/torch/_inductor/ir.py#L1704-L1731)) since `x` is a `SliceView` and that isn't supported.

Thus, I tried adding support for `SliceView` in `as_storage_and_layout`. This worked for my case, but if instead `sizes` had backed symints e.g. `sizes=[s0, 256]` then some benchmarked models lost accuracy.
```
    if isinstance(x, SliceView):
        return as_storage_and_layout(
            x.data,
            freeze=freeze,
            want_contiguous=want_contiguous,
            stride_order=stride_order,
        )
```

So instead of the above, I tried unwrapping the `SliceView` via `x = x.unwrap_view()`. This works for my usecase and passes CI but I'm not entirely sure why. If we unwrap our `SliceView` and create a `ReinterpretView`, I'd assume we'd lose the reindexer from `SliceView`. ~~But maybe we can re-create the same indexing from the `ReinterpretView`'s strides?~~ edit: we do lose vital information (like offset) when you release your `SliceView` and create a `ReinterpretView` so that's a no-go.

Moving onto the final version of this PR. We call `ExternKernel.realize_input()` (feels a bit weird to use `ExternKernel` but it's exactly what I need). It will go ahead and handle our `SliceView` case ([see](https://github.com/pytorch/pytorch/blob/a468b9fbdf75b2da128e5fac4284ebbb981cd31a/torch/_inductor/ir.py#L3733-L3739)) by converting it to a `ReinterpretView` with the correct offset.


# Test
```
$ python test/inductor/test_unbacked_symints.py
..
----------------------------------------------------------------------
Ran 10 tests in 20.813s

OK
```




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ColinPeppler committed:
Update on "[inductor] For View.create(x, sizes) call realize_input() instead of realize() when handling unbacked symints"

# Context
Let's say we do `View.create(x, sizes)` where `x` is a `SliceView` and `sizes` contains unbacked symints e.g. `sizes = [i14, 256]`. Then, this we'll run ([this code](https://github.com/pytorch/pytorch/blob/7e37f63e5e77a217a08d9261af49e63ddb51d306/torch/_inductor/ir.py#L2058-L2071)) where we.
1. Call `x.realize()` -- SliceView(Pointwise) -> SliceView(ComputedBuffer).
2. Retrieve storage & layout via `as_storage_and_layout(x)`
3. Calculate `new_layout` based off layout & `new_sizes`
3. `return ReinterpretView(storage, new_layout)`
However, (2) will raise `NotImplementedError` ([see](https://github.com/pytorch/pytorch/blob/7e37f63e5e77a217a08d9261af49e63ddb51d306/torch/_inductor/ir.py#L1704-L1731)) since `x` is a `SliceView` and that isn't supported.

Thus, I tried adding support for `SliceView` in `as_storage_and_layout`. This worked for my case, but if instead `sizes` had backed symints e.g. `sizes=[s0, 256]` then some benchmarked models lost accuracy.
```
    if isinstance(x, SliceView):
        return as_storage_and_layout(
            x.data,
            freeze=freeze,
            want_contiguous=want_contiguous,
            stride_order=stride_order,
        )
```

So instead of the above, I tried unwrapping the `SliceView` via `x = x.unwrap_view()`. This works for my usecase and passes CI but I'm not entirely sure why. If we unwrap our `SliceView` and create a `ReinterpretView`, I'd assume we'd lose the reindexer from `SliceView`. ~~But maybe we can re-create the same indexing from the `ReinterpretView`'s strides?~~ edit: we do lose vital information (like offset) when you release your `SliceView` and create a `ReinterpretView` so that's a no-go.

Moving onto the final version of this PR. We call `ExternKernel.realize_input()` (feels a bit weird to use `ExternKernel` but it's exactly what I need). It will go ahead and handle our `SliceView` case ([see](https://github.com/pytorch/pytorch/blob/a468b9fbdf75b2da128e5fac4284ebbb981cd31a/torch/_inductor/ir.py#L3733-L3739)) by converting it to a `ReinterpretView` with the correct offset.


# Test
```
$ python test/inductor/test_unbacked_symints.py
..
----------------------------------------------------------------------
Ran 10 tests in 20.813s

OK
```




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ColinPeppler commented:
I was seeing `AssertionError: Redefinition of test inner_device_cuda` in CI tests. So, hopefully using `instantiate_device_type_tests()` instead of `instantiate_parametrized_tests()` resolves it.
